test/unit/app/utils/test_ussd.py

```diff
@@ -0,0 +1,61 @@
+import pytest
+
+from server import db
+from fixtures.user import UserFactory
+from server.models.ussd import UssdMenu, UssdSession
+
+
+@pytest.fixture(scope='function')
+def menu_display_text_in_lang(test_client, init_database):
+    from server.utils.ussd.ussd import menu_display_text_in_lang
+    return menu_display_text_in_lang
+
+
+@pytest.fixture(scope='function')
+def create_or_update_session(test_client, init_database):
+    from server.utils.ussd.ussd import create_or_update_session
+    return create_or_update_session
+
+
+# TODO(ussd): once preferred_language is added, add another test that sw_KE works and user without preferred_language
+@pytest.mark.parametrize("case,expected", [
+    (1, "foo"),
+    (2, "foo"),
+])
+def test_menu_display_text_in_lang(menu_display_text_in_lang, case, expected):
+    if case == 1:
+        user = None
+    else:
+        # TODO: this is really gross, but I can't use the factory in the parameters since no db initiated... how do?
```

Comment:
ideally it'd be:

```
@pytest.mark.parametrize("user,expected", [
    (None, "foo"),
    (UserFactory(), "foo"),
])
```

the error i'd get is:

```
RuntimeError: No application found. Either work inside a view function or push an application context. See http://flask-sqlalchemy.pocoo.org/contexts/
```

Reply:
this is also really gross because i'm passing `init_database` into the `menu_display_text_in_lang` fixture, even though that method doesn't at all need the database, just the factory in the other part of the test does

Reply:
mmmmmm gonna have to understand the ussd stuff better before i can comment properly on this i think

Reply:
hmmm i can't think of an answer currently, i think this reflects a broader problem we have with db objects inside tests

Reply:
more "mmmmmm". It looks like pytest necessarily evaluates fixtures during collection, so u need some way of force-deferring the evaluation. The only way I could force defer while keeping all parameters in the parameterize decorator was to use a partial, but it's not necessarily any cleaner.

`from functools import partial
# TODO(ussd): once preferred_language is added, add another test that sw_KE works and user without preferred_language
@pytest.mark.parametrize("user_factory,expected", [
    (None, "foo"),
    (partial(UserFactory,first_name="Bob"), "foo"),
])
def test_menu_display_text_in_lang(test_client, init_database, user_factory, expected):
    user = user_factory() if user_factory else None

    menu = UssdMenu(display_text_en="foo", display_text_sw="bar")
    assert menu_display_text_in_lang(menu, user) == expected
`

Reply:
is pytest the standard for doing testing in python? maybe it's just me lacking familiarity with python's paradigms, but in both js and ruby tests, you'd spin up a db instance before _all_ tests are run, then each test or test suite can define their own spinning up and down of db objects, and then finally after all tests finish it closes the db. i can see the benefits of spinning up db only when needed within each test, but that doesn't seem efficient for performance / usability

Reply:
if that is how things are just done, your approach looks much cleaner